src/generators/notebookgenerator/itemrenderer.tsx

```diff
@@ -13,8 +15,161 @@ import { sanitizerOptions } from '../shared';
 
+/**
+ * Returns the header level for a provided cell.
+ *
+ * ## Notes
+ *
+ * -   If a cell does not contain a header, the function returns the sentinel value `-1`.
+ *
+ * -   Cell header examples:
+ *
+ *     -   Markdown header:
+ *
+ *         ```
+ *         # Foo
+ *         ```
+ *
+ *     -   Markdown header:
+ *
+ *         ```
+ *         Foo
+ *         ===
+ *         ```
+ *
+ *         ```
+ *         Foo
+ *         ---
+ *         ```
+ *
+ *     -   HTML heading:
+ *
+ *         ```
+ *         <h3>Foo</h3>
+ *         ```
+ *
+ * @private
+ * @param cell - notebook cell
+ * @returns header level
+ */
+function headerLevel(cell: Cell): number {
+  if (cell.constructor.name !== 'MarkdownCell') {
+    return -1;
+  }
+  const lines = cell.model.value.text.split('\n');
+
+  // Case: Markdown header
+  let match = lines[0].match(/^([#]{1,6}) (.*)/);
+  if (match) {
+    return match[1].length;
+  }
+  // Case: Markdown header
+  if (lines.length > 1) {
+    match = lines[1].match(/^([=]{2,}|[-]{2,})/);
+    if (match) {
+      return match[1][0] === '=' ? 1 : 2;
+    }
+  }
+  // Case: HTML heading (WARNING: this is not particularly robust, as HTML headings can span multiple lines)
+  match = lines[0].match(/<h([1-6])>(.*)<\/h\1>/i);
+  if (match) {
+    return parseInt(match[1], 10);
+  }
+  return -1;
+}
+
+/**
+ * Collapses a notebook cell by hiding its section-defined sub-cells (i.e., cells which have lower precedence).
+ *
+ * @private
+ * @param tracker - notebook tracker
+ * @param cell - notebook cell
+ */
+function collapseCell(tracker: any, cell: Cell): void {
+  // Guard against attempting to collapse already hidden cells...
+  if (cell.isHidden) {
+    return;
+  }
+  const level: number = headerLevel(cell);
+
+  // Guard against attempting to collapse cells which are not "collapsible" (i.e., do not define sections)...
+  if (level === -1) {
+    return;
+  }
+  const widgets = tracker.currentWidget.content.widgets;
+  const len = widgets.length;
+  const idx = widgets.indexOf(cell);
+
+  // Guard against an unrecognized "cell" argument...
+  if (idx === -1) {
+    return;
+  }
+  // Search for notebook cells which are semantically defined as sub-cells...
+  for (let i = idx + 1; i < len; i++) {
+    let w = widgets[i];
+    let l: number = headerLevel(w);
+
+    // Check if a widget is at the same or higher level...
+    if (l >= 0 && l <= level) {
+      // We've reached the end of the section...
+      break;
+    }
+    // Collapse a sub-cell by setting its `hidden` state:
+    w.setHidden(true);
+  }
+  // Set a meta-data flag to indicate that we've collapsed notebook sections:
+  cell.model.metadata.set('toc-nb-collapsed', true);
+}
+
+/**
+ * Expands a notebook cell by displaying its section-defined sub-cells (i.e., cells which have lower precedence).
+ *
+ * @private
+ * @param tracker - notebook tracker
+ * @param cell - notebook cell
+ */
+function uncollapseCell(tracker: any, cell: Cell): void {
```

Comment:
This function is very similar to the above. Maybe they can be generalized to something like `setCollapseState` with a boolean parameter to collapse/uncollapse?

Reply:
Done.